File: training task/abstraction.py
from abc import ABC, abstractmethod

class Student(ABC):
    @abstractmethod
    def display_details(self):
        pass

class UndergraduateStudent(Student):
    def __init__(self, name, year):
        self.name = name
        self.year = year

    def display_details(self):
        print("Undergraduate Student:")
        print(f"Name: {self.name}")
        print(f"Year: {self.year}")

class GraduateStudent(Student):
    def __init__(self, name, thesis_title):
        self.name = name
        self.thesis_title = thesis_title

    def display_details(self):
        print("Graduate Student:")
        print(f"Name: {self.name}")
        print(f"Thesis Title: {self.thesis_title}")


undergrad = UndergraduateStudent("Muni", 3)
grad = GraduateStudent("Nazim", "Machine Learning in Healthcare")

undergrad.display_details()
grad.display_details()

File: training task/abstraction.cpp
#include <iostream>
using namespace std;

// Abstract base class
class Student {
public:
    virtual void displayDetails() = 0; // do nothing function-->Pure virtual function
};

class UndergraduateStudent : public Student {
private:
    string name;
    int year;

public:
    UndergraduateStudent( string name, int year) : name(name) {
        this->year=year;
    }

    void displayDetails() override {
        cout << "Undergraduate Student:" << endl;
        cout << "Name: " << name << endl;
        cout << "Year: " << year << endl;
    }
};

class GraduateStudent : public Student {
private:
    string name;
    string thesisTitle;

public:
    GraduateStudent(const string& name, const string& thesisTitle) : name(name), thesisTitle(thesisTitle) {}

    void displayDetails() override {
        cout << "Graduate Student:" << endl;
        cout << "Name: " << name << endl;
        cout << "Thesis Title: " << thesisTitle << endl;
    }
};

int main() {
    UndergraduateStudent undergrad("Alice", 3);
    GraduateStudent grad("Bob", "Machine Learning in Healthcare");
    grad.displayDetails();

    return 0;
}
